### PR TITLE
karmada-search: ResourceRegistry controller can't handle DeletedFinalStateUnknown it gets from controller framework

### PR DESCRIPTION
**Pull request by xigang**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Fix panic in SearchController caused by receiving DeletedFinalStateUnknown object from the cache.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```
Fix panic in ResourceRegistry controller caused by receiving DeletedFinalStateUnknown object from the cache.
```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
Hi @xigang, thanks for your feedback.

Can you explain a little bit when a panic occurs? I wonder if we need to do this in other places for the delete event.

BTW, can you help update the release note.

**Comment by xigang** [comment]
> Hi @xigang, thanks for your feedback.
> 
> Can you explain a little bit when a panic occurs? I wonder if we need to do this in other places for the delete event.
> 
> BTW, can you help update the release note.

@XiShanYongYe-Chang `DeleteFunc` is called when an existing resource is deleted. It gets the final state of the resource (if it is known). Otherwise, it gets an object of type DeletedFinalStateUnknown. This can happen if the watch is closed and misses the delete event and the controller doesn't notice the deletion until the subsequent re-list.

https://github.com/kubernetes/client-go/blob/master/tools/cache/delta_fifo.go#L800

**Comment by XiShanYongYe-Chang** [comment]
@xigang Thanks for your explanation.

Ask @RainbowMango to help rerun the test, the ci failed for `setup environment` failed, it's no related with the updated code.

File: pkg/search/controller.go [hunk 1]
```diff
@@ -331,7 +331,22 @@ func (c *Controller) updateResourceRegistry(oldObj, newObj interface{}) {
 
 // deleteResourceRegistry parse the resourceRegistry object and add deleted Cluster to the queue
 func (c *Controller) deleteResourceRegistry(obj interface{}) {
-	rr := obj.(*searchv1alpha1.ResourceRegistry)
+	rr, isRR := obj.(*searchv1alpha1.ResourceRegistry)
+	// We can get DeletedFinalStateUnknown instead of *clusterV1alpha1.Cluster here and
```
Comment (XiShanYongYe-Chang) [inline pkg/search/controller.go:335]: clusterV1alpha1.Cluster -> searchv1alpha1.ResourceRegistry

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Hi @xigang, do you think we need to cherry-pick this patch to the previous branch?

File: pkg/search/controller.go (hunk 1 above)
Comment (xigang) [reply pkg/search/controller.go]: @XiShanYongYe-Chang The code comment has been fixed.

I think this patch need to cherry-pick the previous branch, and need to check the other controller `cache.ResourceEventHandlerFuncs` `DeleteFunc` dealt with the problem.

Comment (XiShanYongYe-Chang) [reply pkg/search/controller.go]: > I think this patch need to cherry-pick the previous branch and need to check whether other controller of `DeleteFunc` handled `DeletedFinalStateUnknown` types of object.

Thanks~
Can you help do it?

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/lgtm
/approve

**Comment by XiShanYongYe-Chang** [comment]
Can you help add the component name to the release note, just like that:
```
karmada-search: xxx
```

File: pkg/search/controller.go (hunk 1 above)
Comment (xigang) [reply pkg/search/controller.go]: > > I think this patch need to cherry-pick the previous branch and need to check whether other controller of `DeleteFunc` handled `DeletedFinalStateUnknown` types of object.
> 
> Thanks~ Can you help do it?

Okay, I will deal with this issue later.

**Comment by xigang** [comment]
> Can you help add the component name to the release note, just like that:
> 
> ```
> karmada-search: xxx
> ```

done

**Comment by XiShanYongYe-Chang** [comment]
Thanks for your quick response.

**Pull request merged**